### PR TITLE
Disable IPv6 in the default installation

### PR DESCRIPTION
**Pull request by tgraf**
This standardizes the enablement option to `--enable-ipv4` and `--enable-ipv6` while preserving the old `--disable-ipv4` for backward compatibility.

It changes both the Docker and CNI plugin to support with IPv6 mode disabled. The plugins will refuse to provide service if neither address is provided.

It makes all IPv6 code in the datapath optional so it can be disabled.

Finally, the PR disabled IPv6 for all new deployments while preserving the existing behavior for existing deployments. This reduces the memory footprint in default installations significantly.

Depends on: #6080
Fixes: #1537 

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6082)

### CODE REVIEW

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

```
19:12:00 VBoxManage: error: VMDK: cannot write allocated data block in '/root/VirtualBox VMs/ubuntu-18.04.1-amd64-126_1543691411260_68257/ubuntu-18.04.1-amd64-126-disk001.vmdk' (VERR_DISK_FULL)

```

**Comment by tgraf** [comment]
test-me-please

```
21:26:07 There was an error while executing `VBoxManage`, a CLI used by Vagrant
21:26:07 for controlling VirtualBox. The command and stderr is shown below.
21:26:07 
21:26:07 Command: ["snapshot", "5c03a1e1-4218-48c6-887c-1d88725449b7", "list", "--machinereadable"]
21:26:07 
21:26:07 Stderr: VBoxManage: error: The object functionality is limited
21:26:07 VBoxManage: error: Details: code E_ACCESSDENIED (0x80070005), component MachineWrap, interface IMachine, callee nsISupports
21:26:07 VBoxManage: error: Context: "LockMachine(a->session, LockType_Shared)" at line 331 of file VBoxManageSnapshot.cpp
```

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

Runtime Flake:
`[Fail] RuntimePolicyImportTests [It] checks policy trace output `

k8s tests passed

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-missed-k8s

File: Documentation/cmdref/cilium-agent.md [hunk 1]
```diff
@@ -37,9 +37,10 @@ cilium-agent
   -d, --device string                               Device facing cluster/external network for direct L3 (non-overlay mode) (default "undefined")
       --disable-conntrack                           Disable connection tracking
       --disable-endpoint-crd                        Disable use of CiliumEndpoint CRD
-      --disable-ipv4                                Disable IPv4 mode
       --disable-k8s-services                        Disable east-west K8s load balancing by cilium
   -e, --docker string                               Path to docker runtime socket (DEPRECATED: use container-runtime-endpoint instead) (default "unix:///var/run/docker.sock")
+      --enable-ipv4                                 Enable IPv4 support (default true)
```
Comment (joestringer) [inline Documentation/cmdref/cilium-agent.md:42]: (commit title: `Univeral` -> `Universal`)

File: bpf/bpf_lxc.c [hunk 2]
```diff
@@ -707,15 +712,19 @@ int handle_ingress(struct __sk_buff *skb)
...
 		ret = DROP_MISSED_TAIL_CALL;
 		break;
+#endif
 
+#ifdef ENABLE_IPV4
 	case bpf_htons(ETH_P_IP):
 		ep_tail_call(skb, CILIUM_CALL_IPV4_FROM_LXC);
 		ret = DROP_MISSED_TAIL_CALL;
 		break;
+#endif
```
Comment (joestringer) [inline bpf/bpf_lxc.c]: Consider extending the `#endif` to include ARP, given that ARP is used for IPv4 neighbour resolution?

File: daemon/daemon.go [hunk 3]
```diff
@@ -591,15 +597,17 @@ func (d *Daemon) init() error {
...
 
-			if err := lbmap.Service6Map.DeleteAll(); err != nil {
-				return err
+			if option.Config.EnableIPv6 {
+				if err := lbmap.Service6Map.DeleteAll(); err != nil {
+					return err
+				}
+				if err := lbmap.RRSeq6Map.DeleteAll(); err != nil {
+					return err
+				}
```
Comment (joestringer) [inline daemon/daemon.go:622]: Does it make sense for the cleanup to be specific to whether IPv4/IPv6 is enabled? This cleanup functionality is in relation to a previous Cilium run, and we don't know whether the previous run had IPv4/IPv6 enabled, so if someone deploys Cilium with standard DS then disables eg IPv6 then the maps will never be cleared?

File: daemon/state.go [hunk 4]
```diff
@@ -325,22 +325,24 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState) {
...
 		if err != nil {
-			ipam.ReleaseIP(ep.IPv6.IP())
+			// TODO if allocation failed reallocate a new IP address and setup veth
+			// pair accordingly
+			return fmt.Errorf("unable to reallocate IPv6 address: %s", err)
 		}
-	}(ep)
+
+		defer func(ep *endpoint.Endpoint) {
+			if err != nil {
```
Comment (joestringer) [inline daemon/state.go:337]: For this defer function to work as intended, the same `err` variable needs to be shared across the entire function.

File: pkg/ipcache/ipcache.go [hunk 5]
```diff
@@ -288,7 +288,7 @@ func (ipc *IPCache) Upsert(ip string, hostIP net.IP, newIdentity Identity) bool
 			}
 		}
 	} else {
-		scopedLog.Error("Attempt to upsert invalid IP into ipcache layer")
+		scopedLog.Panic("Attempt to upsert invalid IP into ipcache layer")
```
Comment (joestringer) [inline pkg/ipcache/ipcache.go]: This change in log level is surprising given that it's causing a panic in a commit that doesn't explicitly call this out. Is this intentional?

**Review by joestringer — DISMISSED** [review]
Looks good overall.

A couple of notes:
* The deferred release of allocated IPv6 address below doesn't look like it works correctly in error conditions
* It'd be nice to teach `cilium-health` that the IPv4 and/or IPv6 addresses are not expected to work, so that it doesn't always complain about lack of connectivity. (Could be a follow-up)

File: daemon/daemon.go (hunk 3 above)
Comment (joestringer) [reply daemon/daemon.go:622]: Looks like the final commit takes care of this concern.

**Comment by joestringer** [comment]
Is there any additional route manipulation that is missing the IPv4/IPv6 wrappers?

**Comment by tgraf** [comment]
Found another bug, when switching from IPv6-disabled to IPv6-enabled, endpoints are not being restored because of:
```
Dec 04 23:37:03 k8s1 cilium-agent[20127]: level=error msg="Failed to re-allocate IP of endpoint. Not restoring endpoint." endpointID=37533 error="unable to reallocate IPv6 address: provided IP is not in the valid range. The range of valid IPs is f00d::a10:0:0:0/112" subsys=daemon
```

File: pkg/ipcache/ipcache.go (hunk 5 above)
Comment (tgraf) [reply pkg/ipcache/ipcache.go]: Good catch, this was a leftover from debugging.

**Comment by tgraf** [comment]
> Is there any additional route manipulation that is missing the IPv4/IPv6 wrappers?

Not that I found. Most of the route code doing node routes depends on the IP address of the address family being set or not. I ran with both IPv4 and IPv6 disabled and did not experience warnings about routes being added unnecessarily.

There is a bit of a chicken-egg problem when a user disables a protocol for an existing cluster. The endpoints will keep using the address family until the pod is restarted whereas the routes will already be gone. I'm not sure how to solve this better. It's what I would expect as a user anyway.

**Comment by tgraf** [comment]
test-me-please

memcached runtime tests have failed
```
 runtime.RuntimeMemcache Testing binary memcache Tests basic memcache operation
```

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-missed-k8s

**Comment by tgraf** [comment]
test-missed-k8s

**Comment by tgraf** [comment]
The long-lived connection tests while restarting are failing. Likely related to the change.

File: daemon/state.go (hunk 4 above)
Comment (aanm) [reply daemon/state.go:337]: @tgraf @joestringer that's not enough, the returned error needs to have the variable `err` assigned.

```
func (d *Daemon) allocateIPsLocked(ep *endpoint.Endpoint) (err error) {
....
```

File: daemon/state.go [hunk 6]
```diff
@@ -325,26 +325,27 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState) {
...
-	err := ipam.AllocateIP(ep.IPv6.IP())
-	if err != nil {
-		// TODO if allocation failed reallocate a new IP address and setup veth
-		// pair accordingly
-		return fmt.Errorf("unable to reallocate IPv6 address: %s", err)
-	}
+	var err error
 
-	defer func(ep *endpoint.Endpoint) {
+	if option.Config.EnableIPv6 && ep.IPv6 != nil {
```
Comment (aanm) [inline daemon/state.go:330]: If the endpoint didn't contain an IPv6 shouldn't we try to assign a new one now?

**Review by aanm — DISMISSED** [review]

File: daemon/state.go (hunk 6 above)
Comment (tgraf) [reply daemon/state.go:330]: That's potentially quite dangerous as the application inside the container is already running. Assigning an IPv6 address in the middle may break things. Just like we keep using addresses when disabling an address family, we should probably not add a new address to already existing containers.

**Comment by aanm** [comment]
@tgraf forgot to mention, might worth to check if both enable-ipv4 and enable-ipv6 are not set to false before starting the daemon.

**Review by raybejjani — DISMISSED** [review]
Looks good except for the `allocateIPsLocked` err handling.

**Comment by tgraf** [comment]
> forgot to mention, might worth to check if both enable-ipv4 and enable-ipv6 are not set to false before starting the daemon.

Done

File: daemon/state.go (hunk 4 above)
Comment (tgraf) [reply daemon/state.go:337]: @aanm Why? I don't think that is true at all.

**Comment by tgraf** [comment]
test-me-please

File: daemon/state.go (hunk 4 above)
Comment (aanm) [reply daemon/state.go:337]: @tgraf https://play.golang.org/p/IjBet0lGfY9

Comment (tgraf) [reply daemon/state.go:337]: I see what you mean but we only need err to be non-nil which will be the case. The defer to clean up is called when necessary.

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-missed-k8s

**Comment by tgraf** [comment]
test-missed-k8s

**Comment by tgraf** [comment]
test-missed-k8s

**Comment by tgraf** [comment]
test-missed-k8s

**Review by aanm — APPROVED** [review]
I'm glad the default IPv6 is still `true`

**Pull request merged**